cmd/key_generate.go

```diff
@@ -2,183 +2,175 @@ package cmd
-		Long:  "Generate a new key pair for a wallet, this will implicitly generate a new wallet if none exist for the given wallet",
-		RunE:  runKeyGenerate,
-	}
-)
+	generateKeyLong = cli.LongDesc(`
+		Generate a new Ed25519 key pair in a given wallet.
 
-func init() {
-	keyCmd.AddCommand(keyGenerateCmd)
-	keyGenerateCmd.Flags().StringVarP(&keyGenerateArgs.wallet, "wallet", "w", "", "Name of the wallet to use")
-	keyGenerateCmd.Flags().StringVarP(&keyGenerateArgs.passphraseFile, "passphrase-file", "p", "", "Path of the file containing the passphrase to access the wallet")
-	keyGenerateCmd.Flags().StringVarP(&keyGenerateArgs.metas, "meta", "m", "", `A list of metadata e.g: "primary:true;asset:BTC"`)
-	_ = keyGenerateCmd.MarkFlagRequired("wallet")
-}
-
-func runKeyGenerate(_ *cobra.Command, _ []string) error {
-	p := printer.NewHumanPrinter()
+		If the targeted wallet doesn't exist, it will be automatically generated.
+	`)
 
-	store, err := wallets.InitialiseStore(rootArgs.home)
-	if err != nil {
-		return fmt.Errorf("couldn't initialise wallets store: %w", err)
-	}
+	generateKeyExample = cli.Examples(`
+		# Generate a key pair
+		vegawallet key generate --wallet WALLET
 
-	handler := wallets.NewHandler(store)
-
-	walletExists := handler.WalletExists(keyGenerateArgs.wallet)
-
-	passphrase, err := getPassphrase(keyGenerateArgs.passphraseFile, !walletExists)
-	if err != nil {
-		return err
-	}
+		# Generate a key pair with additional metadata (name = my-wallet and type = validation)
+		vegawallet key generate --wallet WALLET --meta "name:my-wallet,type:validation"
 
-	metas, err := parseMeta(keyGenerateArgs.metas)
-	if err != nil {
-		return err
-	}
+		# Generate a key pair with custom name
+		vegawallet key generate --wallet WALLET --meta "name:my-wallet"
+	`)
+)
 
-	var mnemonic string
-	if !walletExists {
-		if rootArgs.output == "human" {
-			p.BangMark().Text("Wallet ").Bold(keyGenerateArgs.wallet).Text(" does not exist yet").NextLine()
-		}
+type GenerateKeyHandler func(flags.PassphraseGetterWithOps, *wallet.GenerateKeyRequest) (*wallet.GenerateKeyResponse, error)
 
-		mnemonic, err = handler.CreateWallet(keyGenerateArgs.wallet, passphrase)
+func NewCmdGenerateKey(w io.Writer, rf *RootFlags) *cobra.Command {
+	h := func(passphraseGetter flags.PassphraseGetterWithOps, req *wallet.GenerateKeyRequest) (*wallet.GenerateKeyResponse, error) {
+		s, err := wallets.InitialiseStore(rf.Home)
 		if err != nil {
-			return fmt.Errorf("couldn't create wallet: %w", err)
+			return nil, fmt.Errorf("couldn't initialise wallets store: %w", err)
 		}
 
-		if rootArgs.output == "human" {
-			p.CheckMark().Text("Wallet ").Bold(keyGenerateArgs.wallet).Text(" has been created at: ").SuccessText(store.GetWalletPath(keyGenerateArgs.wallet)).NextLine()
+		// Because the passphrase needs to be retrieved based on the state of the
+		// specified wallet, we get it here and fill the request with it.
+		passphrase, err := passphraseGetter(s.WalletExists(req.Wallet))
```

Comment:
Late to the party: Was just having a play around, and I think this should be `!s.WalletExists(req.Wallet)`? We want the confirmation when the wallet is created and not every subsequent time we use the wallet.

Reply:
🦅 👁️